### PR TITLE
environmentd: assert tracing spans

### PR DESCRIPTION
**Pull request by mjibson**
Add a capturing tracing layer during tests that can be used to assert various things about spans. Use this to verify that expected finishing spans are created during some queries. This is useful because we have at many times in the past broken tracing during refactors.

### Motivation

  * This PR adds a feature that has not yet been specified.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - n/a

### CODE REVIEW

**Comment by mjibson** [comment]
This took much more plumbing than I wanted, but this is what I could get to compile. I tried to make the layers in ore/tracing.rs generic over any `Layer` but ran into unreadable rust errors and decided to hard code it to the fluent layer instead.

**Review by ParkMyCar — APPROVED** [review]
This looks awesome!

File: src/environmentd/tests/server.rs [hunk 1]
```diff
@@ -3541,3 +3542,44 @@ async fn test_connection_id() {
     let pid_envid = pid >> 19;
     assert_eq!(envid_lower, pid_envid);
 }
+
+// Test that expected spans are generated for various queries.
+#[test] // allow(test-attribute)
+#[cfg_attr(miri, ignore)] // too slow
+fn test_expected_spans() {
+    // TODO: When trying to make this test async, it blocks forever while trying to drop `server`.
```
[the remark this answers is not in the document]
Comment (guswynn) [reply src/environmentd/tests/server.rs:3550]: sounds like something is blocking a tokio task somewhere...

**Review by guswynn — APPROVED** [review]
@mjibson this is sweet!!!

**Comment by guswynn** [comment]
> I tried to make the layers in ore/tracing.rs generic over any Layer but ran into unreadable rust errors and decided to hard code it to the fluent layer instead.

this is possible but VERY annoying, hard-coding seems way better!

**Comment by mjibson** [comment]
I've now spent many hours trying to figure out why the test isn't passing. It passes locally and locally using mzcompose, so I'm pretty baffled. Might abandon this and come back later.

**Comment by guswynn** [comment]
@mjibson Perhaps it needs to be in its own integration test file?

**Comment by mjibson** [comment]
Same error when that test is moved to its own integration test.

**Comment by guswynn** [comment]
@mjibson I'll have to take a look at this...

**Comment by mjibson** [comment]
Idea: this is maybe a race or a difference between debug and release builds. Changed the test to explicitly drop the client then run the asserts in a Retry.

**Comment by guswynn** [comment]
@mjibson and I determined that only calls that call `create_collections` seem to fail; the failures reproduce locally on my laptop. we should merge with those tests disabled

**Comment by mjibson** [comment]
Merged as part of #26083.

**Pull request closed without merging**